### ISSUE TITLE
Add slicing to distributions

### ISSUE THREAD

**Issue by Rizhiy** [pytorch/pytorch#52625]
## 🚀 Feature
When `Distribution` is batched, it should support slicing, returning new `Distribution` with appropriate parameters.

## Motivation
I need to separate batched distributions in my code, now I have to write custom code for each class.

## Pitch

e.g. for `Normal` distribution:
```python
from torch.distributions.normal import Normal
import torch

a = Normal(torch.zeros(32), torch.ones(32))

a_1 = a[:16]
a_1 == Normal(torch.zeros(16), torch.ones(16))
```

## Alternatives

N/A

## Additional context

N/A

**Comment by fritzo** [comment]
This should be pretty easy in 1.8, now that the `.arg_constraints` have `.event_dim` attributes. Ignoring the complexities of advanced indexing for a moment, I believe we should be able to do something generic that will work for most Distribution subclasses:
```py
class Distribution:
    def __getitem__(self, index):
        if not isinstance(index, tuple):
            index = (index,)
        params = {}
        colon = (slice(None),)
        for name, constraint in self.arg_constraints.items():
            params[name] = getattr(self, name)[index + colon * constraint.event_dim]
        result = type(self)(**params, validate_args=False)
        result._validate_args = getattr(self, "_validate_args", None)
        return result
```
Then we'd need a few special cases for e.g. `LKJCholesky` and some distributions like `Categorical` that can take either probs or logits...

**Comment by mmcdermott** [comment]
Any progress on this? It would be a valuable feature.
